solver/progress.go

```diff
@@ -230,52 +234,55 @@ func (p *progressUI) Release() {
 	close(p.done)
 }
 
-func (p *progressUI) Wait() error {
-	return p.g.Wait()
-}
-
-func NewDebugProgress(ctx context.Context) Progress {
-	g, ctx := errgroup.WithContext(ctx)
-
-	done := make(chan struct{})
-	g.Go(func() error {
-		<-done
-		return nil
-	})
+func (p *progressUI) Wait() (err error) {
+	defer p.spp.cancel()
+	defer func() {
+		gerr := p.g.Wait()
```

Comment:
Isn't this redundant with the Wait at the end of the function? Also, will this ever be called in a way that would race with the p.g assignment?

Reply:
Thanks, fixed the extra `p.g.Wait` at the end, forgot to delete it after the refactor.  As for a race, yeah maybe, I have modified it so `Sync` calls `waitNoLock` since the `Sync` already holds the lock, and then changed  `Wait` to acquire a lock then call `waitNoLock`

Reply:
I think this would be a bit more readable without the second `defer`:

```go
func (p *progressUI) waitNoLock() error {
	defer p.spp.cancel()
	<-p.done
	err := p.spp.wait()
	gerr := p.g.Wait()
	if err == nil {
		return gerr
	}
	return err
}
```

Reply:
updated